{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2907d8ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np \n",
    "import csv \n",
    "from Bio import SeqIO\n",
    "from itertools import islice \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "82f3148f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('gi|221721649|gb|CP000628.1|', 4005130)\n",
      "('gi|221725460|gb|CP000629.1|', 2650913)\n",
      "('gi|221727862|gb|CP000630.1|', 184668)\n",
      "('gi|221728079|gb|CP000631.1|', 388169)\n",
      "('gi|221728482|gb|CP000632.1|', 44420)\n"
     ]
    }
   ],
   "source": [
    "#Reads in Fasta reference file and stores information in a list using SeqIO\n",
    "#Use information about the length of first and second chromosome to determine early and late mutations \n",
    "\n",
    "def referenceStats(pathToReferenceFasta):\n",
    "    \n",
    "    #Creates a list of records for each sequence in the fasta file \n",
    "    records = list(SeqIO.parse(pathToReferenceFasta, \"fasta\"))\n",
    "    \n",
    "    #Loop through the record objects and yield the chromosome name and length of the chromosome \n",
    "    for num in range(len(records)):\n",
    "        sample_len = (len(records[num].seq))\n",
    "        yield (records[num].name, sample_len)\n",
    "    \n",
    "# Test to see if function works \n",
    "genObj = referenceStats(\"/Users/jadetakakuwa/Desktop/SungLab/currentData/Radio/rradio_copy.fa\")\n",
    "for item in genObj: \n",
    "    print(item)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d7865fb8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4005129\n"
     ]
    }
   ],
   "source": [
    "#List of origins and termini according to Cd Viz for each organism \n",
    "rradioOrigin = 496696 \n",
    "rradioTerm = 2499261\n",
    "vitisOrigin = 336131 \n",
    "vitisTerm = 2199319 \n",
    "tumOrigin = 2765000\n",
    "tumTerm = 1344210\n",
    "\n",
    "#Function to adjust postion according to origin and terminus in Cd Viz \n",
    "#Shifts the origin given by Cd Viz to 1 and adjusts other positions accordingly \n",
    "def adjustPosition(pathToReferenceFasta, position, origin):\n",
    "    \n",
    "    #If position is less than the origin, subtract from the full length of the chromosome \n",
    "    if position < origin: \n",
    "        genObj = referenceStats(pathToReferenceFasta)\n",
    "        for i in islice(genObj, 0, 1):\n",
    "            fullchromLength = i[1]  \n",
    "        adjPosition = fullchromLength - (origin -(position + 1))\n",
    "        \n",
    "    #if position is greater than the origin, subtract origin from position and add 1 since the origin is at position 1 \n",
    "    elif position > origin: \n",
    "        adjPosition = (position - origin) + 1  \n",
    "        \n",
    "    #Readjust origin to position 1 \n",
    "    elif position == origin: \n",
    "        adjPosition = 1 \n",
    "    \n",
    "    else: \n",
    "        adjPosition = \"No position\"\n",
    "    \n",
    "    return adjPosition \n",
    "\n",
    "\n",
    "#Test\n",
    "# print(adjustPosition(\"/Users/jadetakakuwa/Desktop/SungLab/currentData/Radio/rradio_copy.fa\",496695, rradioOrigin))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "57b52e43",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample_name': 'C10', 1763764: 'GC_TA', 3649628: 'GC_AT', 3649632: 'AT_TA', 3649633: 'AT_TA', 1763773: 'GC_AT', 1700377: 'AT_TA', 3468115: 'GC_AT', 3468112: 'AT_TA', 3468105: 'GC_AT'}\n",
      "{'Sample_name': 'A6', 3414082: 'AT_CG', 916848: 'GC_TA', 1334795: 'AT_GC', 1802022: 'GC_AT'}\n",
      "{'Sample_name': 'B3', 1763743: 'AT_TA', 2821955: 'GC_AT', 1763780: 'GC_TA', 1763773: 'GC_AT', 1763764: 'GC_TA', 1763757: 'AT_TA', 3649633: 'AT_GC', 3649632: 'AT_TA', 3649608: 'AT_GC', 3649604: 'AT_GC', 3649600: 'GC_TA', 3649598: 'AT_CG', 3649594: 'AT_GC'}\n",
      "{'Sample_name': 'B11', 32834: 'GC_AT', 1855904: 'AT_GC'}\n",
      "{'Sample_name': 'A4', 1445019: 'GC_TA'}\n",
      "{'Sample_name': 'B6', 3547482: 'GC_AT'}\n",
      "{'Sample_name': 'C9', 309078: 'GC_AT', 2761226: 'GC_AT', 953688: 'GC_AT', 664656: 'GC_CG', 3476734: 'GC_AT'}\n",
      "{'Sample_name': 'D6', 3411906: 'AT_CG', 3038214: 'AT_TA', 375872: 'GC_TA', 1622319: 'GC_AT'}\n",
      "{'Sample_name': 'D2', 779064: 'GC_TA'}\n",
      "{'Sample_name': 'B2', 310768: 'GC_CG', 591827: 'AT_GC', 3765708: 'GC_CG'}\n",
      "{'Sample_name': 'C5', 310626: 'AT_CG', 13304: 'GC_TA', 648433: 'GC_AT', 1518046: 'GC_AT', 3257203: 'GC_AT', 1248132: 'GC_CG'}\n",
      "{'Sample_name': 'D7', 1766287: 'GC_AT', 2899847: 'GC_AT'}\n",
      "{'Sample_name': 'D9', 3699159: 'GC_TA'}\n",
      "{'Sample_name': 'B10', 1289381: 'GC_TA', 3602032: 'GC_TA', 1403208: 'AT_CG', 1403280: 'GC_TA', 3065975: 'GC_AT'}\n",
      "{'Sample_name': 'C7'}\n",
      "{'Sample_name': 'A2', 1517170: 'AT_CG', 1263301: 'AT_GC', 378973: 'GC_AT'}\n",
      "{'Sample_name': 'A3'}\n",
      "{'Sample_name': 'C2', 3699159: 'GC_TA'}\n",
      "{'Sample_name': 'B9'}\n",
      "{'Sample_name': 'B12'}\n",
      "{'Sample_name': 'C6'}\n",
      "{'Sample_name': 'D5', 3190550: 'GC_CG', 3237078: 'GC_AT', 3302600: 'AT_GC'}\n",
      "{'Sample_name': 'D3', 1607159: 'GC_CG', 1873171: 'GC_TA'}\n",
      "{'Sample_name': 'B5', 1542572: 'GC_TA', 1486821: 'GC_AT', 1472895: 'GC_AT', 1153082: 'GC_AT', 2792345: 'GC_AT'}\n",
      "{'Sample_name': 'C8', 2245352: 'AT_GC', 913575: 'AT_GC'}\n",
      "{'Sample_name': 'B7', 2451590: 'GC_TA', 1261779: 'GC_TA'}\n",
      "{'Sample_name': 'D8', 834146: 'AT_GC', 1768496: 'AT_GC'}\n",
      "{'Sample_name': 'A10', 1971894: 'GC_CG'}\n",
      "{'Sample_name': 'A5', 1768496: 'AT_GC', 834146: 'AT_GC', 740426: 'GC_AT'}\n",
      "{'Sample_name': 'A1', 2684058: 'GC_AT', 2421840: 'AT_GC'}\n",
      "{'Sample_name': 'A8', 3673662: 'GC_TA', 2844868: 'GC_AT'}\n",
      "{'Sample_name': 'D1', 2692963: 'GC_AT', 926562: 'GC_TA'}\n"
     ]
    }
   ],
   "source": [
    "#Calculate Transitions and transversions\n",
    "#Goes through an excel file with multiple samples of VCF calls\n",
    "#Note on formatting of excel file: keeps columns of the VCF, but must have a Sample Name column to distinguish the sample\n",
    "#Generates new dictionary for each sample with positions of each variant call serving as keys\n",
    "#Values are the sample name and mutation type\n",
    "#Adjusts position according to origin and terminus in Cd Viz \n",
    "\n",
    "def transitionsTransversions(pathToExcelFile, pathToReferenceFasta, origin):\n",
    "    \n",
    "    #Reads in all sheets in an excel file and stores as a dictionary \n",
    "    excelTotalSheets = pd.read_excel(pathToExcelFile, sheet_name=None)\n",
    "    \n",
    "    #Get keys in the dictionary \n",
    "    keysList = excelTotalSheets.keys()\n",
    "    \n",
    "    #Loop through each sample in the excel file \n",
    "    for sample in keysList: \n",
    "        count = 0 \n",
    "        value = pd.DataFrame(excelTotalSheets[str(sample)])\n",
    "        sampleDict = {}\n",
    "        while count < len(value.index):\n",
    "            sampleName = value.iloc[count][\"Sample_name\"]\n",
    "            sampleDict[\"Sample_name\"] = sampleName\n",
    "            position = value.iloc[count][\"POS\"]\n",
    "            adjPos = adjustPosition(pathToReferenceFasta, position, origin)\n",
    "            ref = value.iloc[count][\"REF\"]\n",
    "            alt = value.iloc[count][\"ALT\"]\n",
    "            \n",
    "            if ref == \"A\" and alt == \"G\": \n",
    "                sampleDict[adjPos] = \"AT_GC\"\n",
    "                count += 1 \n",
    "            elif ref == \"T\" and alt == \"C\":\n",
    "                sampleDict[adjPos] = \"AT_GC\"\n",
    "                count += 1\n",
    "            elif ref == \"G\" and alt == \"A\":\n",
    "                sampleDict[adjPos] = \"GC_AT\"\n",
    "                count += 1\n",
    "            elif ref == \"C\" and alt == \"T\":\n",
    "                sampleDict[adjPos] = \"GC_AT\"\n",
    "                count += 1\n",
    "            elif ref == \"A\" and alt == \"T\": \n",
    "                sampleDict[adjPos] = \"AT_TA\"\n",
    "                count += 1\n",
    "            elif ref == \"T\" and alt == \"A\": \n",
    "                sampleDict[adjPos] = \"AT_TA\"\n",
    "                count += 1\n",
    "            elif ref == \"G\" and alt == \"T\":\n",
    "                sampleDict[adjPos] = \"GC_TA\"\n",
    "                count += 1\n",
    "            elif ref == \"C\" and alt == \"A\":\n",
    "                sampleDict[adjPos] = \"GC_TA\"\n",
    "                count += 1\n",
    "            elif ref == \"A\" and alt == \"C\":\n",
    "                sampleDict[adjPos] = \"AT_CG\"\n",
    "                count += 1\n",
    "            elif ref == \"T\" and alt == \"G\":\n",
    "                sampleDict[adjPos] = \"AT_CG\"\n",
    "                count += 1\n",
    "            elif ref == \"G\" and alt == \"C\":\n",
    "                sampleDict[adjPos] = \"GC_CG\"\n",
    "                count += 1\n",
    "            elif ref == \"C\" and alt == \"G\":\n",
    "                sampleDict[adjPos] = \"GC_CG\"\n",
    "                count += 1\n",
    "            else: \n",
    "                count += 1\n",
    "        yield sampleDict\n",
    "\n",
    "# Test \n",
    "# for i in transitionsTransversions(\"/Users/jadetakakuwa/Desktop/SungLab/currentData/Radio/radio_new_vcf_2.xlsx\", \"/Users/jadetakakuwa/Desktop/SungLab/currentData/Radio/rradio_copy.fa\", rradioOrigin):\n",
    "#     print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "89cd0161",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Categorize parts of the chromosome as early or late according to the possible scenarios \n",
    "\n",
    "#Use the generator object from the referenceStats function \n",
    "#Assumptions: the first and second chromosomes are the first and second generator objects, make sure this is true! \n",
    "\n",
    "#Set early and late mutation positions according to Scenario 1\n",
    "#Early replicating is half the length of chromosome 2 divided in half around the origin of chromosome 1\n",
    "#Both forks fire at the same time on both chromosomes \n",
    "#So length of chromosome 2 divided by 4. \n",
    "def scenario1(pathToReferenceFasta): \n",
    "    genObj = referenceStats(pathToReferenceFasta)\n",
    "    \n",
    "    for i in islice(genObj, 0, 1):\n",
    "        chrom1Len = i[1]\n",
    "    \n",
    "    for i in islice(genObj, 0, 1):\n",
    "        chrom2Len = i[1]\n",
    "\n",
    "    halfLength = int(chrom2Len/4)\n",
    "    afterOrigin = halfLength \n",
    "    beforeOrigin = chrom1Len - halfLength \n",
    "    \n",
    "    #Early covers position 1 to afterOrigin, and beforeOrigin to the full length of the chromosome\n",
    "    #Late is every position other than early \n",
    "    return(afterOrigin, beforeOrigin)\n",
    "\n",
    "#Set up early and late according to Scenario 2\n",
    "#Early replicating is before chromosome 2 begins replicating. \n",
    "#The late replicating section is half the length of chromosome 2 on either side of the terminus.\n",
    "#Replication on the first chromosome starts first, then both chromosomes finish replicating at the same time \n",
    "def scenario2(pathToReferenceFasta):\n",
    "    genObj = referenceStats(pathToReferenceFasta)\n",
    "    \n",
    "    for i in islice(genObj, 0, 1):\n",
    "        chrom1Len = i[1]\n",
    "    \n",
    "    for i in islice(genObj, 0, 1):\n",
    "        chrom2Len = i[1]\n",
    "        \n",
    "    halfLength = int(chrom2Len/4)\n",
    "    terminus = int(chrom1Len/2)\n",
    "    afterTerm = terminus + halfLength \n",
    "    beforeTerm = terminus - halfLength \n",
    "    \n",
    "    #Early covers any position before late \n",
    "    #Late covers positions from beforeTerm to afterTerm and includes the terminus \n",
    "    return(beforeTerm, afterTerm)\n",
    "\n",
    "#Test \n",
    "# print(scenario1(\"/Users/jadetakakuwa/Desktop/SungLab/currentData/Tumefaciens/Agrobacterium_tumefaciens_full_wplasmids.fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "37f41b77",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculate number of total AT and CG sites according to early and late replication positions \n",
    "#Returns total AT and CG sites for early mutation positions and total AT and CG sites for late mutation positions \n",
    "#Parameter of function is either scenario 1 or scenario 2 \n",
    "def earlyLateConditional(scenario, pathToReferenceFasta): \n",
    "    genObj = referenceStats(pathToReferenceFasta)\n",
    "    for i in islice(genObj, 0, 1):\n",
    "        chrom1Seq = i[0].seq\n",
    "        chrom1Len = i[1]\n",
    "        if scenario == \"scenario1\":\n",
    "            afterOrigin, beforeOrigin = scenario1(pathToReferenceFasta)\n",
    "            \n",
    "            AT_sites_early = chrom1Seq[0:afterOrigin].count(\"A\") + chrom1Seq[0:afterOrigin].count(\"T\") + \\\n",
    "            chrom1Seq[beforeOrigin:chrom1Len].count(\"A\") + chrom1Seq[beforeOrigin:chrom1Len].count(\"T\")\n",
    "            \n",
    "            CG_sites_early = chrom1Seq[0:afterOrigin].count(\"C\") + chrom1Seq[0:afterOrigin].count(\"G\") + \\\n",
    "            chrom1Seq[beforeOrigin:chrom1Len].count(\"C\") + chrom1Seq[beforeOrigin:chrom1Len].count(\"G\")\n",
    "            \n",
    "            AT_sites_late = chrom1Seq[afterOrigin:beforeOrigin].count(\"A\") + \\\n",
    "            chrom1Seq[afterOrigin:beforeOrigin].count(\"T\") \n",
    "            \n",
    "            CG_sites_late = chrom1Seq[afterOrigin:beforeOrigin].count(\"C\") + \\\n",
    "            chrom1Seq[afterOrigin:beforeOrigin].count(\"G\") \n",
    "            \n",
    "            return(AT_sites_early, CG_sites_early, AT_sites_late, CG_sites_late)\n",
    "        \n",
    "        elif scenario == \"scenario2\":\n",
    "            beforeTerm, afterTerm = scenario2(pathToReferenceFasta)\n",
    "            \n",
    "            AT_sites_early = chrom1Seq[0:beforeTerm].count(\"A\") + chrom1Seq[0:beforeTerm].count(\"T\") + \\\n",
    "            chrom1Seq[afterTerm:chrom1Len].count(\"A\") + chrom1Seq[afterTerm:chrom1Len].count(\"T\")\n",
    "            \n",
    "            AT_sites_late = chrom1Seq[beforeTerm:afterTerm].count(\"A\") + chrom1Seq[beforeTerm:afterTerm].count(\"T\")\n",
    "            \n",
    "            CG_sites_early = chrom1Seq[0:beforeTerm].count(\"C\") + chrom1Seq[0:beforeTerm].count(\"G\") + \\\n",
    "            chrom1Seq[afterTerm:chrom1Len].count(\"C\") + chrom1Seq[afterTerm:chrom1Len].count(\"G\")\n",
    "            \n",
    "            CG_sites_late = chrom1Seq[beforeTerm:afterTerm].count(\"C\") + chrom1Seq[beforeTerm:afterTerm].count(\"G\")\n",
    "            \n",
    "            return(AT_sites_early, CG_sites_early, AT_sites_late, CG_sites_late)\n",
    "\n",
    "#Test     \n",
    "#print(earlyLateConditional(\"scenario2\", \"/Users/jadetakakuwa/Desktop/SungLab/currentData/Tumefaciens/Agrobacterium_tumefaciens_full_wplasmids.fasta\"))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "71bea4c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Early': ['GC_AT', 'AT_TA', 'AT_TA', 'GC_AT', 'AT_TA', 'GC_AT'], 'Late': ['GC_TA', 'GC_AT', 'AT_TA'], 'Sample_name': 'C10'}\n",
      "{'Early': ['AT_CG'], 'Late': ['GC_TA', 'AT_GC', 'GC_AT'], 'Sample_name': 'A6'}\n",
      "{'Early': ['AT_GC', 'AT_TA', 'AT_GC', 'AT_GC', 'GC_TA', 'AT_CG', 'AT_GC'], 'Late': ['AT_TA', 'GC_AT', 'GC_TA', 'GC_AT', 'GC_TA', 'AT_TA'], 'Sample_name': 'B3'}\n",
      "{'Early': ['GC_AT'], 'Late': ['AT_GC'], 'Sample_name': 'B11'}\n",
      "{'Early': [], 'Late': ['GC_TA'], 'Sample_name': 'A4'}\n",
      "{'Early': ['GC_AT'], 'Late': [], 'Sample_name': 'B6'}\n",
      "{'Early': ['GC_AT', 'GC_AT'], 'Late': ['GC_AT', 'GC_AT', 'GC_CG'], 'Sample_name': 'C9'}\n",
      "{'Early': ['AT_CG', 'GC_TA'], 'Late': ['AT_TA', 'GC_AT'], 'Sample_name': 'D6'}\n",
      "{'Early': [], 'Late': ['GC_TA'], 'Sample_name': 'D2'}\n",
      "{'Early': ['GC_CG', 'AT_GC', 'GC_CG'], 'Late': [], 'Sample_name': 'B2'}\n",
      "{'Early': ['AT_CG', 'GC_TA', 'GC_AT'], 'Late': ['GC_AT', 'GC_AT', 'GC_CG'], 'Sample_name': 'C5'}\n",
      "{'Early': [], 'Late': ['GC_AT', 'GC_AT'], 'Sample_name': 'D7'}\n",
      "{'Early': ['GC_TA'], 'Late': [], 'Sample_name': 'D9'}\n",
      "{'Early': ['GC_TA'], 'Late': ['GC_TA', 'AT_CG', 'GC_TA', 'GC_AT'], 'Sample_name': 'B10'}\n",
      "{'Early': [], 'Late': [], 'Sample_name': 'C7'}\n",
      "{'Early': ['GC_AT'], 'Late': ['AT_CG', 'AT_GC'], 'Sample_name': 'A2'}\n",
      "{'Early': [], 'Late': [], 'Sample_name': 'A3'}\n",
      "{'Early': ['GC_TA'], 'Late': [], 'Sample_name': 'C2'}\n",
      "{'Early': [], 'Late': [], 'Sample_name': 'B9'}\n",
      "{'Early': [], 'Late': [], 'Sample_name': 'B12'}\n",
      "{'Early': [], 'Late': [], 'Sample_name': 'C6'}\n",
      "{'Early': [], 'Late': ['GC_CG', 'GC_AT', 'AT_GC'], 'Sample_name': 'D5'}\n",
      "{'Early': [], 'Late': ['GC_CG', 'GC_TA'], 'Sample_name': 'D3'}\n",
      "{'Early': [], 'Late': ['GC_TA', 'GC_AT', 'GC_AT', 'GC_AT', 'GC_AT'], 'Sample_name': 'B5'}\n",
      "{'Early': [], 'Late': ['AT_GC', 'AT_GC'], 'Sample_name': 'C8'}\n",
      "{'Early': [], 'Late': ['GC_TA', 'GC_TA'], 'Sample_name': 'B7'}\n",
      "{'Early': [], 'Late': ['AT_GC', 'AT_GC'], 'Sample_name': 'D8'}\n",
      "{'Early': [], 'Late': ['GC_CG'], 'Sample_name': 'A10'}\n",
      "{'Early': [], 'Late': ['AT_GC', 'AT_GC', 'GC_AT'], 'Sample_name': 'A5'}\n",
      "{'Early': [], 'Late': ['GC_AT', 'AT_GC'], 'Sample_name': 'A1'}\n",
      "{'Early': ['GC_TA'], 'Late': ['GC_AT'], 'Sample_name': 'A8'}\n",
      "{'Early': [], 'Late': ['GC_AT', 'GC_TA'], 'Sample_name': 'D1'}\n"
     ]
    }
   ],
   "source": [
    "#Categorize transitions and transversions from the VCF/excel file as early or late mutations as well as mutation type \n",
    "#Stores information in a dictionary with lists corresponding to early, late and the sample name \n",
    "#Specify scenario 1 or 2 to output the correct dictionary \n",
    "\n",
    "def categorizeMutations(pathToExcelFile, pathToReferenceFasta, origin, scenario):\n",
    "    if scenario == \"scenario1\":  \n",
    "\n",
    "        afterOrigin, beforeOrigin = scenario1(pathToReferenceFasta)\n",
    "\n",
    "        for i in transitionsTransversions(pathToExcelFile, pathToReferenceFasta, origin):\n",
    "            sampleDict = {\"Early\":[], \"Late\":[]}\n",
    "            for key in i.keys():\n",
    "                if key == \"Sample_name\":\n",
    "                    sampleDict.update({\"Sample_name\":i[key]})\n",
    "                elif key != \"Sample_name\":\n",
    "                    if key >= 1 and key <= afterOrigin: \n",
    "                        sampleDict[\"Early\"].append(i[key])\n",
    "                    elif int(key) >= beforeOrigin:\n",
    "                        sampleDict[\"Early\"].append(i[key])\n",
    "                    else: \n",
    "                        sampleDict[\"Late\"].append(i[key])\n",
    "                        \n",
    "            yield sampleDict\n",
    "        \n",
    "    elif scenario == \"scenario2\":\n",
    "        \n",
    "        beforeTerm, afterTerm = scenario2(pathToReferenceFasta)\n",
    "        \n",
    "        for i in transitionsTransversions(pathToExcelFile, pathToReferenceFasta, origin):\n",
    "            sampleDict = {\"Early\":[], \"Late\":[]}\n",
    "            for key in i.keys():\n",
    "                if key == \"Sample_name\":\n",
    "                    sampleDict.update({\"Sample_name\":i[key]})\n",
    "                elif key != \"Sample_name\":\n",
    "                    if key >= beforeTerm and key <= afterTerm: \n",
    "                        sampleDict[\"Late\"].append(i[key])\n",
    "                    elif key < beforeTerm or key > afterTerm: \n",
    "                        sampleDict[\"Early\"].append(i[key])\n",
    "                        \n",
    "            yield sampleDict\n",
    "                               \n",
    "#Test       \n",
    "for i in categorizeMutations(\"/Users/jadetakakuwa/Desktop/SungLab/currentData/Radio/radio_new_vcf_2.xlsx\", \"/Users/jadetakakuwa/Desktop/SungLab/currentData/Radio/rradio_copy.fa\", rradioOrigin, \"scenario1\"):\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "0be16dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Generation times for all organisms \n",
    "generationTimeRadio = 2492\n",
    "generationTimeVitis = 2512 \n",
    "generationTimeTum = 5819\n",
    "\n",
    "#Input sample name, mutation type and whether the mutation is classified as early or late \n",
    "#Calculates the conditional mutation rate and adjusts according to generation time \n",
    "def calculateConditionalRates(sampleDict, pathToReferenceFasta, generationTime, scenario):\n",
    "    \n",
    "    earlyDict = {\"AT_GC\": 0, \"GC_AT\": 0, \"AT_TA\": 0, \"GC_TA\": 0, \"AT_CG\": 0, \"GC_CG\": 0, \"Sample_name\": \"\"}\n",
    "    lateDict = {\"AT_GC\": 0, \"GC_AT\": 0, \"AT_TA\": 0, \"GC_TA\": 0, \"AT_CG\": 0, \"GC_CG\": 0, \"Sample_name\": \"\"}\n",
    "    \n",
    "    if scenario == \"scenario1\": \n",
    "        AT_sites_early, CG_sites_early, AT_sites_late, CG_sites_late = earlyLateConditional(\"scenario1\", pathToReferenceFasta)\n",
    "        for key in sampleDict.keys():\n",
    "            if key == \"Early\":\n",
    "                for mutation in sampleDict[key]:\n",
    "                    if mutation == \"AT_GC\":\n",
    "                        earlyDict[\"AT_GC\"] += 1\n",
    "                    elif mutation == \"GC_AT\":\n",
    "                        earlyDict[\"GC_AT\"] += 1\n",
    "                    elif mutation == \"AT_TA\":\n",
    "                        earlyDict[\"AT_TA\"] += 1\n",
    "                    elif mutation == \"GC_TA\":\n",
    "                        earlyDict[\"GC_TA\"] += 1\n",
    "                    elif mutation == \"AT_CG\":\n",
    "                        earlyDict[\"AT_CG\"] += 1\n",
    "                    elif mutation == \"GC_CG\": \n",
    "                        earlyDict[\"GC_CG\"] += 1 \n",
    "                    else: \n",
    "                        pass \n",
    "            elif key == \"Late\":\n",
    "                for mutation in sampleDict[key]:\n",
    "                    if mutation == \"AT_GC\":\n",
    "                        lateDict[\"AT_GC\"] += 1\n",
    "                    elif mutation == \"GC_AT\":\n",
    "                        lateDict[\"GC_AT\"] += 1\n",
    "                    elif mutation == \"AT_TA\":\n",
    "                        lateDict[\"AT_TA\"] += 1\n",
    "                    elif mutation == \"GC_TA\":\n",
    "                        lateDict[\"GC_TA\"] += 1\n",
    "                    elif mutation == \"AT_CG\":\n",
    "                        lateDict[\"AT_CG\"] += 1\n",
    "                    elif mutation == \"GC_CG\": \n",
    "                        lateDict[\"GC_CG\"] += 1 \n",
    "                    else: \n",
    "                        pass \n",
    "            elif key == \"Sample_name\":\n",
    "                earlyDict[\"Sample_name\"] = sampleDict[key] \n",
    "                lateDict[\"Sample_name\"] = sampleDict[key]\n",
    "\n",
    "    \n",
    "        for key in earlyDict.keys():\n",
    "            if key == \"AT_GC\" or key == \"AT_TA\" or key == \"AT_CG\":\n",
    "                mutRate = earlyDict[key]/AT_sites_early\n",
    "                mutRate = mutRate/generationTime\n",
    "                earlyDict[key] = mutRate\n",
    "            elif key == \"GC_AT\" or key == \"GC_TA\" or key == \"GC_CG\":\n",
    "                mutRate = earlyDict[key]/CG_sites_early\n",
    "                mutRate = mutRate/generationTime\n",
    "                earlyDict[key] = mutRate \n",
    "\n",
    "        for key in lateDict.keys():\n",
    "            if key == \"AT_GC\" or key == \"AT_TA\" or key == \"AT_CG\":\n",
    "                mutRate = lateDict[key]/AT_sites_late\n",
    "                mutRate = mutRate/generationTime\n",
    "                lateDict[key] = mutRate\n",
    "            elif key == \"GC_AT\" or key == \"GC_TA\" or key == \"GC_CG\":\n",
    "                mutRate = lateDict[key]/CG_sites_late\n",
    "                mutRate = mutRate/generationTime\n",
    "                lateDict[key] = mutRate \n",
    "                \n",
    "        return earlyDict, lateDict\n",
    "    \n",
    "    elif scenario == \"scenario2\": \n",
    "        AT_sites_early, CG_sites_early, AT_sites_late, CG_sites_late = earlyLateConditional(\"scenario2\", pathToReferenceFasta)\n",
    "        for key in sampleDict.keys():\n",
    "            if key == \"Early\":\n",
    "                for mutation in sampleDict[key]:\n",
    "                    if mutation == \"AT_GC\":\n",
    "                        earlyDict[\"AT_GC\"] += 1\n",
    "                    elif mutation == \"GC_AT\":\n",
    "                        earlyDict[\"GC_AT\"] += 1\n",
    "                    elif mutation == \"AT_TA\":\n",
    "                        earlyDict[\"AT_TA\"] += 1\n",
    "                    elif mutation == \"GC_TA\":\n",
    "                        earlyDict[\"GC_TA\"] += 1\n",
    "                    elif mutation == \"AT_CG\":\n",
    "                        earlyDict[\"AT_CG\"] += 1\n",
    "                    elif mutation == \"GC_CG\": \n",
    "                        earlyDict[\"GC_CG\"] += 1 \n",
    "                    else: \n",
    "                        pass \n",
    "            elif key == \"Late\":\n",
    "                for mutation in sampleDict[key]:\n",
    "                    if mutation == \"AT_GC\":\n",
    "                        lateDict[\"AT_GC\"] += 1\n",
    "                    elif mutation == \"GC_AT\":\n",
    "                        lateDict[\"GC_AT\"] += 1\n",
    "                    elif mutation == \"AT_TA\":\n",
    "                        lateDict[\"AT_TA\"] += 1\n",
    "                    elif mutation == \"GC_TA\":\n",
    "                        lateDict[\"GC_TA\"] += 1\n",
    "                    elif mutation == \"AT_CG\":\n",
    "                        lateDict[\"AT_CG\"] += 1\n",
    "                    elif mutation == \"GC_CG\": \n",
    "                        lateDict[\"GC_CG\"] += 1\n",
    "                    else: \n",
    "                        pass \n",
    "            elif key == \"Sample_name\":\n",
    "                earlyDict[\"Sample_name\"] = sampleDict[key] \n",
    "                lateDict[\"Sample_name\"] = sampleDict[key]\n",
    "\n",
    "    \n",
    "        for key in earlyDict.keys():\n",
    "            if key == \"AT_GC\" or key == \"AT_TA\" or key == \"AT_CG\":\n",
    "                mutRate = earlyDict[key]/AT_sites_early\n",
    "                mutRate = mutRate/generationTime\n",
    "                earlyDict[key] = mutRate\n",
    "            elif key == \"GC_AT\" or key == \"GC_TA\" or key == \"GC_CG\":\n",
    "                mutRate = earlyDict[key]/CG_sites_early\n",
    "                mutRate = mutRate/generationTime\n",
    "                earlyDict[key] = mutRate \n",
    "\n",
    "        for key in lateDict.keys():\n",
    "            if key == \"AT_GC\" or key == \"AT_TA\" or key == \"AT_CG\":\n",
    "                mutRate = lateDict[key]/AT_sites_late\n",
    "                mutRate = mutRate/generationTime\n",
    "                lateDict[key] = mutRate\n",
    "            elif key == \"GC_AT\" or key == \"GC_TA\" or key == \"GC_CG\":\n",
    "                mutRate = lateDict[key]/CG_sites_late\n",
    "                mutRate = mutRate/generationTime\n",
    "                lateDict[key] = mutRate \n",
    "\n",
    "    return earlyDict, lateDict\n",
    "\n",
    "#Create empty dataframes for early and late mutations \n",
    "final_mutations_early = pd.DataFrame(columns=[\"AT_GC\", \"GC_AT\", \"AT_TA\", \"GC_TA\", \"AT_CG\", \"GC_CG\", \"Sample_name\"])\n",
    "final_mutations_late = pd.DataFrame(columns=[\"AT_GC\", \"GC_AT\", \"AT_TA\", \"GC_TA\", \"AT_CG\", \"GC_CG\", \"Sample_name\"])\n",
    "\n",
    "#For each sample, create a dictionary for early and late mutations\n",
    "#Add the dictionaries to the early and late mutation dataframes \n",
    "for i in categorizeMutations(\"/Users/jadetakakuwa/Desktop/SungLab/currentData/Tumefaciens/tum_new_vcf_2.xlsx\", \"/Users/jadetakakuwa/Desktop/SungLab/currentData/Tumefaciens/Agrobacterium_tumefaciens_full_wplasmids.fasta\", tumOrigin, \"scenario2\"):\n",
    "    earlyDict, lateDict = calculateConditionalRates(i, \"/Users/jadetakakuwa/Desktop/SungLab/currentData/Tumefaciens/Agrobacterium_tumefaciens_full_wplasmids.fasta\", generationTimeTum, \"scenario2\")\n",
    "    final_mutations_early = final_mutations_early.append(earlyDict, ignore_index=True)\n",
    "    final_mutations_late = final_mutations_late.append(lateDict, ignore_index=True)\n",
    "\n",
    "#Write information to an excel file \n",
    "final_mutations_early.to_excel(\"tum_scenario2_early.xlsx\")\n",
    "final_mutations_late.to_excel(\"tum_scenario2_late.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac664523",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
